### PR TITLE
[SQL] [DO NOT MERGE] Try to log some useful things to catch the cause of SPARK-12089.

### PR DESCRIPTION
**Pull request by yhuai**
This PR adds logs in different place and hopefully we can catch the cause of https://issues.apache.org/jira/browse/SPARK-12089.

### CODE REVIEW

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java [hunk 1]
```diff
@@ -445,7 +450,13 @@ public UTF8String getUTF8String(int ordinal) {
     final long offsetAndSize = getLong(ordinal);
     final int offset = (int) (offsetAndSize >> 32);
     final int size = (int) offsetAndSize;
-    return UTF8String.fromAddress(baseObject, baseOffset + offset, size);
+    final UTF8String str = UTF8String.fromAddress(baseObject, baseOffset + offset, size);
```
Comment (nongli) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java]: also check size >= 0

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/BufferHolder.java [hunk 2]
```diff
@@ -42,7 +47,18 @@ public void grow(int neededSize, UnsafeRow row) {
...
     if (buffer.length < length) {
       // This will not happen frequently, because the buffer is re-used.
-      final byte[] tmp = new byte[length * 2];
+      final byte[] tmp;
+      try {
+        tmp = new byte[length * 2];
+      } catch (NegativeArraySizeException e) {
+        String errorMessage =
+          "NegativeArraySizeException is triggered. The current length is " + buffer.length +
+          ". The new length is " + length + " * 2. totalSize is " + totalSize() +
```
Comment (nongli) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/BufferHolder.java]: * 2 is inside quotes. Probably not what you meant.

Comment (nongli) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/BufferHolder.java]: Sorry read that too fast.

Comment (nongli) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/BufferHolder.java]: Might be helpful to catch the exception here/move current logging to
https://github.com/apache/spark/blob/master/sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UnsafeRowWriter.java#L196

and log some more stuff and the entire row. Is the whole thing corrupt?

Comment (yhuai) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/BufferHolder.java]: Done. https://github.com/apache/spark/pull/10120/files#diff-c654dda59109239bce5498cbc30ab251L196 will dump the unsafe row. https://github.com/apache/spark/pull/10120/files#diff-244a70a91841bbddf6fae17c14c18ce4L378 will dump the row that is human readable.

**Comment by yhuai** [comment]
i am closing it.

**Pull request closed without merging**